waku/v2/protocol/waku_rln_relay/waku_rln_relay_utils.nim

```diff
@@ -952,67 +962,85 @@ proc appendRLNProof*(rlnPeer: WakuRLNRelay, msg: var WakuMessage,
-      rlnPeer.lastSeenMembershipIndex = membershipIndex
+      let lastIndex = members[0].index + members.len.uint - 1
+      let indexGap = lastIndex - rlnPeer.lastSeenmembershipIndex
+      if indexGap > 1.uint:
+        warn "membership index gap, may have lost connection", lastIndex, currIndex=rlnPeer.lastSeenMembershipIndex, indexGap = indexGap
+      rlnPeer.lastSeenMembershipIndex = lastIndex
+      rlnPeer.lastProcessedBlock = blockNumber
+      debug "last processed block", blockNumber = blockNumber
       return ok()
   return handler
 
-proc subscribeToMemberRegistrations(web3: Web3, 
-                                    contractAddress: Address,
-                                    fromBlock: string = "0x0",
-                                    handler: GroupUpdateHandler): Future[Subscription] {.async, gcsafe.} =
-  ## subscribes to member registrations, on a given membership group contract
-  ## `fromBlock` indicates the block number from which the subscription starts
-  ## `handler` is a callback that is called when a new member is registered
-  ## the callback is called with the pubkey and the index of the new member
-  ## TODO: need a similar proc for member deletions
-  var contractObj = web3.contractSender(MembershipContract, contractAddress)
-
-  let onMemberRegistered = proc (pubkey: Uint256, index: Uint256) {.gcsafe.} =
-    debug "onRegister", pubkey = pubkey, index = index
-    var groupUpdateRes: RlnRelayResult[void]
-    try:
-      groupUpdateRes = handler(pubkey, index)
-    except Exception as err:
-      error "failed to handle group update", err = err.msg
-    if groupUpdateRes.isErr():
-      error "Error handling new member registration", err=groupUpdateRes.error()
-
-  let onError = proc (err: CatchableError) =
-    error "Error in subscription", err=err.msg
-
-  return await contractObj.subscribe(MemberRegistered,
-                                     %*{"fromBlock": fromBlock, "address": contractAddress},
-                                     onMemberRegistered,
-                                     onError)
+proc parse*(event: type MemberRegistered, 
+            log: JsonNode): RlnRelayResult[MembershipTuple] =
+  ## parses the `data` parameter of the `MemberRegistered` event
+  ## returns an error if it cannot parse the `data` parameter
+  var pubkey: UInt256
+  var index: UInt256
+  var data: string
+  try:
+    data = strip0xPrefix(log["data"].getStr())
+  except CatchableError as err:
+    return err("failed to parse the data field of the MemberRegistered event: " & err.msg)
+  var offset = 0
+  try:
+    offset += decode(data, offset, pubkey)
+    offset += decode(data, offset, index)
+    return ok((index: index.toMembershipIndex(), 
+               idComm: pubkey.toIDCommitment()))
+  except:
+    return err("failed to parse the data field of the MemberRegistered event")
+
+type BlockTable = OrderedTable[BlockNumber, seq[MembershipTuple]]
+proc getHistoricalEvents*(ethClientUri: string,
```

Comment:
I'm aware some of this looks unsafe. The goal is to get this proc to look like https://github.com/status-im/nimbus-eth2/blob/44f652f704f14c67e648f2b14dee652d44b858c5/ncli/deposit_downloader.nim#L42, which has a cleaner way to handle rate limiting, etc. Will track it in a separate issue, for common web3 utils.